### PR TITLE
bootloader: Let provision.py fail if a public key contains 0xFFFF

### PR DESCRIPTION
**Pull request by osaether**
Added a check in provision.py such that it fails if a public key
contains 0xFFFF.

Ref: NCSDK-5912

Signed-off-by: Ole Sæther <ole.saether@nordicsemi.no>

### CODE REVIEW

File: scripts/bootloader/provision.py [hunk 1]
```diff
@@ -63,7 +68,10 @@ def get_hashes(public_key_files):
     hashes = list()
     for fn in public_key_files:
         with open(fn, 'rb') as f:
-            hashes.append(sha256(VerifyingKey.from_pem(f.read()).to_string()).digest()[:16])
+            key = sha256(VerifyingKey.from_pem(f.read()).to_string()).digest()[:16]
+            if key_contains_ffff(key):
+                raise RuntimeError("Public key containing 0xffff found. Please try again with a valid key")
```
Comment (oyvindronningstad) [inline scripts/bootloader/provision.py:73]: Ideally we should tell the user a) which key triggered this, and b) that it's the hash that has ffff, not the key itself.

**Review by oyvindronningstad — APPROVED** [review]

File: scripts/bootloader/provision.py [hunk 2]
```diff
@@ -63,7 +69,11 @@ def get_hashes(public_key_files):
     hashes = list()
     for fn in public_key_files:
         with open(fn, 'rb') as f:
-            hashes.append(sha256(VerifyingKey.from_pem(f.read()).to_string()).digest()[:16])
+            digest = sha256(VerifyingKey.from_pem(f.read()).to_string()).digest()[:16]
+            if digest_contains_ffff(digest):
+                raise RuntimeError("Hash of key in '%s' contains 0xffff. Please regenerate the key." %
+                                   os.path.abspath(f.name))
```
Comment (tejlmand) [inline scripts/bootloader/provision.py:75]: why not do:
```suggestion
            if digest.find(b'\xff\xff\xff\xff') != -1:
                raise RuntimeError("Hash of key in '%s' contains 0xffff. Please regenerate the key." %
                                   os.path.abspath(f.name))
```

Comment (oyvindronningstad) [reply scripts/bootloader/provision.py:75]: We're looking for half-word aligned 0xFFFFs

Comment (tejlmand) [reply scripts/bootloader/provision.py:75]: right, my suggestion would also fail if the `0xffff` is not half-word aligned.
Question is if that is a real issue. A hash with `0xffff` starting at an odd mem address is just as unlikely as the `0xffff` on a half-word aligned address.

Comment (oyvindronningstad) [reply scripts/bootloader/provision.py:75]: I don't see a problem with the current solution, so why willingly introduce false negatives?

Comment (tejlmand) [reply scripts/bootloader/provision.py:75]: there is nothing wrong.
Just found it nice if we could safe some lines :wink: 

As alternative, one could do:
```suggestion
            if [n for n in range(8) if digest[2*n:2*n+2] == b'\xff\xff']:
```

which gives you what you want, a check on each half-word aligned 2-bytes.

Comment (oyvindronningstad) [reply scripts/bootloader/provision.py:75]: Haha, nice :). maybe like this?

```suggestion
            if any([digest[n:n+2] == b'\xff\xff' for n in range(0, len(digest), 2)]):
                raise RuntimeError("Hash of key in '%s' contains 0xffff. Please regenerate the key." %
                                   os.path.abspath(f.name))
```

Comment (tejlmand) [reply scripts/bootloader/provision.py:75]: I also like that one, but a couple of chars longer (and thus violating PEP a bit more than my version :grin: ).
but both are still shorter than the `raise RuntimeError` line.

Happy to approve both versions.

**Review by oyvindronningstad — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Pull request merged**